cloud/validations.go

```diff
@@ -0,0 +1,256 @@
+// Copyright 2017 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+// Package cloud provides functionality to parse information
+// describing clouds, including regions, supported auth types etc.
+
+package cloud
+
+import (
+	"fmt"
+	"reflect"
+	"strings"
+
+	"github.com/juju/errors"
+	"github.com/juju/gojsonschema"
+	"gopkg.in/yaml.v2"
+)
+
+var cloudSetSchema = map[string]interface{}{
+	"type": "object",
+	"properties": map[string]interface{}{
+		"clouds": map[string]interface{}{
+			"type":                 "object",
+			"additionalProperties": cloudSchema,
+		},
+	},
+	"additionalProperties": false,
+}
+
+var cloudSchema = map[string]interface{}{
+	"type": "object",
+	"properties": map[string]interface{}{
+		"type": map[string]interface{}{"type": "string"},
+		"auth-types": map[string]interface{}{
+			"type":  "array",
+			"items": map[string]interface{}{"type": "string"},
+		},
+		"endpoint":          map[string]interface{}{"type": "string"},
+		"identity-endpoint": map[string]interface{}{"type": "string"},
+		"storage-endpoint":  map[string]interface{}{"type": "string"},
+		"config":            map[string]interface{}{"type": "object"},
+		"regions":           regionsSchema,
+		"region-config":     map[string]interface{}{"type": "object"},
+	},
+	"additionalProperties": false,
+}
+
+var regionsSchema = map[string]interface{}{
+	"type": "object",
+	"additionalProperties": map[string]interface{}{
+		"type": "object",
+		"properties": map[string]interface{}{
+			"endpoint":          map[string]interface{}{"type": "string"},
+			"identity-endpoint": map[string]interface{}{"type": "string"},
+			"storage-endpoint":  map[string]interface{}{"type": "string"},
+		},
+		"additionalProperties": false,
+	},
+}
+
+func validateCloudSet(data []byte) error {
+	var err error
+
+	invalidKeys := validateCloudSetMetaData(data)
+
+	if err != nil {
+		return err
+	}
+
+	formatKeyError := func(invalidKey, similarValidKey string) string {
+		str := fmt.Sprintf("property %s is invalid.", invalidKey)
+		if similarValidKey != "" {
+			str = fmt.Sprintf("%s Perhaps you mean \"%s\".\n", str, similarValidKey)
+		} else {
+			str += "\n"
+		}
+
+		return str
+	}
+
+	msg := ""
+	for k, v := range invalidKeys {
+		msg += formatKeyError(k, v)
+	}
+
+	if msg != "" {
+		err = errors.New(msg)
+	}
+
+	return err
+}
+
+func CloudTags() []string {
```

Comment:
please comment exported functions, types, etc.
(I think this one should be unexported though?)